### PR TITLE
first commit, for code review

### PR DESCRIPTION
**Pull request by sanjeevsiva17**
mail parser using, aws lambda

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "async": "^2.6.1",
+    "aws-sdk": "^2.385.0",
+    "gm": "^1.23.1",
+    "mailparser": "^2.4.3",
+    "mailparser-mit": "^0.6.2",
```
Comment (sudhirj) [inline package.json]: Why are we adding the MIT fork? We want to use the main library, not a fork.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+
+
+exports.handler = function(event, context, callback) {
+    // Read options from the event.
+    var eventBody = event.Records[0].body;
+    var srcBucket = JSON.parse(eventBody).Records[0].s3.bucket.name;
+    // Object key may have spaces or unicode non-ASCII characters.
+    var srcKey    = decodeURIComponent(JSON.parse(eventBody).Records[0].s3.object.key.replace(/\+/g, " "));  
+    var dstBucket = srcBucket + "resized";
```
Comment (sudhirj) [inline index.js]: Take the destination bucket as confit as well, don’t assume anything about the configuration.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    // Sanity check: validate that source and destination are different buckets.
+    if (srcBucket == dstBucket) {
+        callback("Source and destination buckets are the same.");
+        return;
+    }
+
+    //async waterdall docs https://caolan.github.io/async/docs.html#waterfall
+    async.waterfall([
+        function download(next) {
+            // Download the image from S3 into a buffer.
```
Comment (sudhirj) [inline index.js]: What image?

File: index.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            });
+        },
+            
+        function upload(mailattachments, next) {
+            // Stream the transformed image to a different S3 bucket.
+            Promise.all(mailattachments.map((mailattachment) => {
+                return new Promise((resolve, reject) => {
+                    s3.putObject({
+                        Bucket: dstBucket,
+                        Key: mailattachment.filename,
```
Comment (sudhirj) [inline index.js]: Can’t use the file names as keys, if two emails come with attachments with the same file name they’ll ovrewrite each other, no?

File: package.json (hunk 1 above)
Comment (sudhirj) [reply package.json]: Doing the EUPL only because we’re not using the MIT fork.

**Review by sudhirj — CHANGES_REQUESTED** [review]

**Comment by sudhirj** [comment]
Instead of saying first commit, for review, mention what the code actually does. Observe the way PRs are handled on GitHub.com/rails/rails for example.

File: package.json (hunk 1 above)
Comment (sanjeevsiva17) [reply package.json]: Forgot to remove that, I was experimenting

File: index.js (hunk 3 above)
Comment (sanjeevsiva17) [reply index.js]: supposed to be a file, sorry.

File: index.js (hunk 4 above)
Comment (sanjeevsiva17) [reply index.js]: Ok, I will add it in the format 
"emailid"+"epoc"+"filename"

File: index.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+var s3 = new AWS.S3();
+
+
+
+exports.handler = function(event, context, callback) {
+    // Read options from the event.
+    var eventBody = event.Records[0].body;
+    var srcBucket = JSON.parse(eventBody).Records[0].s3.bucket.name;
+    // Object key may have spaces or unicode non-ASCII characters.
+    var srcKey    = decodeURIComponent(JSON.parse(eventBody).Records[0].s3.object.key.replace(/\+/g, " "));
```
Comment (sivachandran) [inline index.js]: you could cache the `s3` object to re-parsing the JSON

File: index.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                contentType = response.ContentType;
+                next(null,mailparse)
+            }).catch((err) => {
+                console.log(err)
+                callback("error parsing attachments.")
+            });
+        },
+            
+        function upload(mailattachments, next) {
+            // Stream the transformed image to a different S3 bucket.
```
Comment (sivachandran) [inline index.js]: irrelevant comment

File: parse.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+const simpleParser = require('mailparser').simpleParser;
+var mailattach = {
+    filename: [],
+    content: []
+}
+
+// mailattch = [{ filename: '', content: '' }, {  }]
+
+module.exports = function mailparser(data){
+    console.log("here");
```
Comment (sivachandran) [inline parse.js]: keep it meaningful if the log the required.

File: index.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            Promise.all(mailattachments.map((mailattachment) => {
+                return new Promise((resolve, reject) => {
+                    s3.putObject({
+                        Bucket: dstBucket,
+                        Key: mailattachment.filename,
+                        Body: mailattachment.content,
+                        ContentType: contentType
```
Comment (sivachandran) [inline index.js:66]: @sudhirj Shouldn't this be hardcoded value like `binary/octet-stream` or shouldn't we inspect the content determine the value?

**Review by sivachandran — CHANGES_REQUESTED** [review]

File: index.js (hunk 8 above)
Comment (sudhirj) [reply index.js:66]: Doesn't really matter what the content type is - will vary wildly on the implementation of the email client / extension. If it looks like a KDM, smells like a KDM, validates as a KDM, (or a zip file) we assume it is.

File: index.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+var mailparse = require('./parse.js');
+// get reference to S3 client 
+var s3 = new AWS.S3();
+
+
+
+exports.handler = function(event, context, callback) {
+    // Read options from the event.
+    var eventBody = event.Records[0].body;
+    sources3 = JSON.parse(eventBody).Records[0].s3;
```
Comment (maknahar) [inline index.js]: let's rename it to sourceS3?

Comment (maknahar) [inline index.js]: @sudhirj @sivachandran Should we be adding the S3 event here? I think routing it via SQS is a better option.

**Review by maknahar — CHANGES_REQUESTED** [review]

File: index.js (hunk 9 above)
Comment (maknahar) [reply index.js]: It seems we are processing the SQS event. We need to delete the message from the queue after processing @sanjeevsiva17.

File: index.js [hunk 10]
```diff
@@ -76,6 +85,13 @@ exports.handler = function(event, context, callback) {
                     ' and uploaded to ' + dstBucket + '/' + dstKey
                 );
             }
+        
+            sqs.deleteMessage(params, function(err, data) {
+                if (err) 
+                    console.log(err, err.stack); // an error occurred
+                else     
+                    console.log("sqs message delete :",data);           // successful response
```
Comment (sivachandran) [inline index.js]: `SQS message deleted`

File: index.js [hunk 11]
```diff
@@ -76,6 +85,13 @@ exports.handler = function(event, context, callback) {
                     ' and uploaded to ' + dstBucket + '/' + dstKey
                 );
             }
+        
+            sqs.deleteMessage(params, function(err, data) {
+                if (err) 
+                    console.log(err, err.stack); // an error occurred
```
Comment (sivachandran) [inline index.js]: add contextual message, `SQS message deletion failed`

File: index.js [hunk 12]
```diff
@@ -11,13 +11,22 @@ var s3 = new AWS.S3();
...
     // Object key may have spaces or unicode non-ASCII characters.
-    var srcKey    = decodeURIComponent(JSON.parse(eventBody).Records[0].s3.object.key.replace(/\+/g, " "));  
-    var dstBucket = srcBucket + "resized";
+    var srcKey    = decodeURIComponent(sourceS3.object.key.replace(/\+/g, " "));  
+    var dstBucket = process.env.S3_BUCKET;
     var dstKey    = srcKey;
 
+    var ReceiptHandle = event.Records[0].ReceiptHandle
+    //to delete queue message
+    var params = {
```
Comment (sivachandran) [inline index.js]: can be named as `deleteMessageParams`

File: parse.js [hunk 13]
```diff
@@ -3,19 +3,26 @@
...
-    console.log("here");
     return new Promise(function(resolve, reject){
         simpleParser(data)
             .then(mail => {
-            resolve(mail.attachments);
+                for (var attach in mail.attachments){
+                mailattach.push({
+                    attachment : mail.attachments[attach],
+                    date : mail.date,
+                    email : regExp.exec(mail.from.text)[1]
```
Comment (sivachandran) [inline parse.js:22]: is it guaranteed that the email always are always enclosed with angle brackets?

File: parse.js [hunk 14]
```diff
@@ -3,19 +3,26 @@
...
-    filename: [],
-    content: []
-}
+var mailattach = []
 
 // mailattch = [{ filename: '', content: '' }, {  }]
+// regex to get email address inside the brackets <>. 
+// Because mail.email returns email: 'Hello world! <helloworld@gmail.com>'
+
+var regExp = /\<([^)]+)\>/;
```
Comment (sivachandran) [inline parse.js]: why `[^)]`? shouldn't it be `[^>]`?

**Review by sivachandran — CHANGES_REQUESTED** [review]

File: parse.js (hunk 13 above)
Comment (sanjeevsiva17) [reply parse.js:22]: I have checked for gmail and yahoo mail. it stores it in the <> brackets. I will look into this !

File: parse.js (hunk 14 above)
Comment (sanjeevsiva17) [reply parse.js]: it should be with the >, will change it

File: index.js (hunk 9 above)
Comment (sanjeevsiva17) [reply index.js]: According to SQS documentation, the message will delete automatically. 
https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html

@sudhirj @sivachandran @maknahar

File: parse.js (hunk 13 above)
Comment (sanjeevsiva17) [reply parse.js:22]: I have edited the regex to accept both angle and square brackets

File: parse.js [hunk 15]
```diff
@@ -5,11 +5,10 @@ const fs = require('fs');
 const simpleParser = require('mailparser').simpleParser;
 var mailattach = []
 
-// mailattch = [{ filename: '', content: '' }, {  }]
-// regex to get email address inside the brackets <>. 
+// regex to get email address from string 
 // Because mail.email returns email: 'Hello world! <helloworld@gmail.com>'
 
-var regExp = /\<([^)]+)\>/;
+var regExp = /([a-zA-Z0-9._-]+@[a-zA-Z0-9._-]+\.[a-zA-Z0-9_-]+)/;
```
Comment (sivachandran) [inline parse.js]: email id can have `+`. I don't think this regex supports it. check https://stackoverflow.com/questions/46155/how-to-validate-an-email-address-in-javascript and https://www.regextester.com/19

**Review by sivachandran — CHANGES_REQUESTED** [review]

File: parse.js (hunk 15 above)
Comment (sanjeevsiva17) [reply parse.js]: https://regex101.com/r/tF2Dqu/1 
i have edited the current regex to accept + sign.

Comment (sudhirj) [reply parse.js]: Why are we validating the email? How does it matter what the email was?

Comment (sudhirj) [reply parse.js]: Ah, I see. Just take whatever is between the brackets, no? Why a regex inside the brackets? Are there cases where the address string contains multiple pieces of text inside angle brackets? Even if that’s the case, can just take the last one until we have contradictory evidence.

**Review by sudhirj — CHANGES_REQUESTED** [review]

File: parse.js (hunk 15 above)
Comment (sivachandran) [reply parse.js]: @sudhirj Not sure the email address is always enclosed with angle/square brackets. So thought of parsing the email address. But now I realise we can craft regex in such way to parse entire string if the angle/square brackets are not there. @sanjeevsiva17

**Comment by sanjeevsiva17** [comment]
Run emailgen.py, to generate eml file, Then run lambda.py to upload the eml file to s3 !

**Comment by grkkarthick90** [comment]
@sanjeevsiva17 Shall we remove `.DS_Store` and `Archive.zip`? I don't see `Archive.zip` is used anywhere in this branch.

File: test/lambda.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+bucket1 = s3.Bucket(bucket_name1)
+print("created destination bucket\n")
+
+fileToBeUploaded = os.getcwd()+'/eml/email_sample.eml'
+key = 'email_sample'
+bucket.upload_file(fileToBeUploaded, key)
+
+for object in bucket1.objects.all():
+  print(object)
```
Comment (grkkarthick90) [inline test/lambda.py]: @sanjeevsiva17 There was proper assertion was not added to ensure the attachment is presented in the destination bucket. Let us create a file name with some uuid and ensure that it is found in the destination bucket.

File: test/lambda.py [hunk 17]
```diff
@@ -0,0 +1,22 @@
+import boto3
+import botocore
+import json
+import os
+
+s3 = boto3.resource('s3', region_name = 'us-east-1')
+
+bucket_name = 'inbox-bucket-test'
```
Comment (grkkarthick90) [inline test/lambda.py]: Rename this to email_blob_bucket

File: test/lambda.py [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+bucket_name1 = 'inbox-bucket-testresized'
```
Comment (grkkarthick90) [inline test/lambda.py]: Rename this to email_attachment_bucket

File: test/lambda.py [hunk 19]
```diff
@@ -0,0 +1,22 @@
+import boto3
```
Comment (grkkarthick90) [inline test/lambda.py]: We need to add a case, email without attachment.

File: test/lambda.py [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+bucket_name = 'inbox-bucket-test'
+bucket_name1 = 'inbox-bucket-testresized'
+
+bucket = s3.Bucket(bucket_name)
+print("created source bucket\n")
+
+bucket1 = s3.Bucket(bucket_name1)
+print("created destination bucket\n")
+
+fileToBeUploaded = os.getcwd()+'/eml/email_sample.eml'
```
Comment (grkkarthick90) [inline test/lambda.py]: You can give the filename, unique `from` address to the script `emailgen.py` and then upload the generated blob to the S3 bucket? So that you can do the assertion on the destination bucket quickly.

File: test/test.py [hunk 21]
```diff
@@ -21,21 +22,29 @@ def upload(filename):
...
   bucket.upload_file(fileToBeUploaded, key)
 
-def assertAttachment(fromaddr, *files):
-  for object in bucket1.objects.all():
-    x = []
-    x.append(object)
-    print(object)
+compare = lambda x, y: collections.Counter(x) == collections.Counter(y)    
+
+def assertAttachment(From, *attachments):
```
Comment (grkkarthick90) [inline test/test.py]: Keep all the local variable names in `lowercase`.

File: test/emailgen.py [hunk 22]
```diff
@@ -0,0 +1,45 @@
+import os
+from email import generator
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+from email.MIMEBase import MIMEBase
+from email import Encoders
+import datetime
+import re 
+
+def Emailgen(From, to, subject, cc, bcc, *attachments):
```
Comment (grkkarthick90) [inline test/emailgen.py]: Keep all the local variables in `lowercase`

File: test/emailgen.py [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+                <body>
+                    <p> hello world </p>
+                </body>
+            </html>
+            """
+
+    msg = MIMEMultipart('alternative')
+    msg['Subject'] = subject
+
+    From = From
```
Comment (grkkarthick90) [inline test/emailgen.py]: What is the use of this statement?

File: test/emailgen.py [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+        Encoders.encode_base64(part)
+        part.add_header('Content-Disposition', 'attachment; filename=' + file)
+        msg.attach(part)
+
+    outfile_name = ("eml/"+From)
+    with open(outfile_name, 'w') as outfile:
+        gen = generator.Generator(outfile)
+        gen.flatten(msg)
+
+    print ("=========== Blob created ============")
```
Comment (grkkarthick90) [inline test/emailgen.py]: We do not need this print statement

File: test/test.py [hunk 25]
```diff
@@ -0,0 +1,61 @@
+import boto3
+import botocore
+import json
+import os
+from emailgen import Emailgen
+import collections
+
+s3 = boto3.resource('s3', region_name = 'us-east-1')
+
+bucket_name = 'inbox-bucket-test'
```
Comment (grkkarthick90) [inline test/test.py]: Change the variable name to `source_bucket`

File: test/test.py [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+bucket_name1 = 'inbox-bucket-testresized'
```
Comment (grkkarthick90) [inline test/test.py]: Change the variable name to `destination_bucket`

File: test/test.py [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+  emailBlob = Emailgen(From, to, Sub, cc, bcc, *attchments)
+  
+  print("Updating bucket information")
+  print("Uploading to bucket\n")
+  upload(From)
+
+  print("Checking Destination bucket\n")
+  assertAttachment(From, *attachments)
+
+From = raw_input("From : ")
```
Comment (grkkarthick90) [inline test/test.py]: This script should not get any input from the user. We can create from address with `uuid` to make it unique for every run.

Comment (grkkarthick90) [reply test/test.py]: Applies to the below variable as well. 
> This script should not get any input from the user

File: test/test.py [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+
+bucket = s3.Bucket(bucket_name)
+print("created source bucket\n")
+
+bucket1 = s3.Bucket(bucket_name1)
+print("created destination bucket\n")
+
+
+def upload(filename):
+  fileToBeUploaded = os.getcwd()+'/eml/'+filename
```
Comment (grkkarthick90) [inline test/test.py]: We should not construct this one as it is already there at https://github.com/RealImage/extraction/pull/1/files#diff-485a45816c4cd61ca20e34c84f5cf564R40.

File: test/test.py [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+    print(objectlist)   
+    result = compare(attachments,objectlist)   
+    if(result == True):
+      print("Succesfully extracted ataachments")
+      print(objectlist)  
+    else:
+      print("Unsucessfull, Check cloudview logs for more details!")
+
+def test1(From, to, Sub, cc, bcc, *attchments):
+  emailBlob = Emailgen(From, to, Sub, cc, bcc, *attchments)
```
Comment (grkkarthick90) [inline test/test.py]: Simply return the file name from the method `Emailgen`

File: test/emailgen.py (hunk 22 above)
Comment (grkkarthick90) [reply test/emailgen.py]: Rename this method name to `GenerateEmailBlob`

File: test/test.py [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+Sub = raw_input("Subject : ")
+cc = raw_input("Cc : ")
+bcc = raw_input("Bcc : ")
+
+number_of_attachments = int(raw_input("Enter the number of attachments: "))
+attachments = []
+for i in range(number_of_attachments):
+  attachments.append(raw_input("Enter name of attachment: "))
+
+test1(From, To, Sub, cc, bcc, *attachments)
```
Comment (grkkarthick90) [inline test/test.py]: We need to add more test like this. We can discuss this in person.

**Review by grkkarthick90 — CHANGES_REQUESTED** [review]
Please do not add .pyc files to the repo.

File: test/emailgen.py (hunk 23 above)
Comment (sanjeevsiva17) [reply test/emailgen.py]: Removed

File: test/test.py (hunk 28 above)
Comment (sanjeevsiva17) [reply test/test.py]: The other one is for attaching the file to the mail, this one is basically to upload the created blob to the bucket

File: test/test.py [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+import uuid 
+import time
+
+s3 = boto3.resource('s3', region_name = 'us-east-1')
+
+source_bucket = 'inbox-bucket-test'
+destination_bucket = 'inbox-bucket-testresized'
+
+bucket = s3.Bucket(source_bucket)
+bucket1 = s3.Bucket(destination_bucket)
```
Comment (grkkarthick90) [inline test/test.py]: Rename this to meaning name

File: test/test.py [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+import collections
+import uuid 
+import time
+
+s3 = boto3.resource('s3', region_name = 'us-east-1')
+
+source_bucket = 'inbox-bucket-test'
+destination_bucket = 'inbox-bucket-testresized'
+
+bucket = s3.Bucket(source_bucket)
```
Comment (grkkarthick90) [inline test/test.py]: Rename this to meaning name

File: test/test.py [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+bucket1 = s3.Bucket(destination_bucket)
+
+def upload(filename, emailblob):
+  fileToBeUploaded = os.getcwd()+ '/' + emailblob
+  key = filename
+  bucket.upload_file(fileToBeUploaded, key)
+
+compare = lambda x, y: collections.Counter(x) == collections.Counter(y)
```
Comment (grkkarthick90) [inline test/test.py]: Move this before `upload` function

File: test/test.py [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+
+def CreateAndUploadEmailBlob(to, Sub, cc, bcc, *attchments):
+  fromadd = str(uuid.uuid4()) + '@gmail.com'
+  print(fromadd)
+  emailBlob = GenerateEmailBlob(fromadd, to, Sub, cc, bcc, *attchments)
+  upload(fromadd, emailBlob)
+  return fromadd
+ 
+
+def test1():
```
Comment (grkkarthick90) [inline test/test.py]: Rename this method name to denote actually what this test does

File: test/test.py [hunk 35]
```diff
@@ -0,0 +1,62 @@
...
+  attachments = ['test.json', 'test.py']
+  fromadd = CreateAndUploadEmailBlob('To@gmail.com', 'Subject', 'cc@yoahoomail.com', 'bcc@ymailcom', *attachments)
+  time.sleep(5)
+  assertAttachment(fromadd, *attachments)
+
```
Comment (grkkarthick90) [inline test/test.py:61]: Add another test to send empty attachment

**Review by grkkarthick90 — CHANGES_REQUESTED** [review]

**Review by grkkarthick90 — APPROVED** [review]
